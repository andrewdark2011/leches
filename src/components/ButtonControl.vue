<script setup lang="ts">
import { computed } from 'vue'
import type { Control } from '../types'

const props = defineProps<{
  label: string
  control: Control
}>()

const classes = computed(() => {
  const base = 'btn'
  const variant = `btn-${props.control.value.variant || 'primary'}`
  const size = `btn-${props.control.value.size || 'sm'}`
  return [base, variant, size]
})
</script>

<template>
  <button
    :class="classes"
    @click="control.value.onClick"
  >
    <i
      class="mr-2"
      :class="control.value.icon"
    />
    {{ control.label }}
  </button>
</template>

<style scoped>
.btn {
  @apply rounded-md cursor-pointer border-none flex items-center;
  @apply font-bold;
  @apply transition-all duration-200;
}

.btn-sm {
  @apply text-xs px-2.5 py-2;
}

.btn-md {
  @apply text-sm px-3 py-2;
}

.btn-lg {
  @apply text-base px-4 py-2;
}

.btn-block {
  @apply px-4 py-2 text-base w-full justify-center;
}

.btn-primary {
  @apply bg-gray-100 text-gray-700;
}
.btn-primary:hover {
  @apply bg-gray-200;
}

.btn-secondary {
  @apply bg-dark-700 text-light-100;
}

.btn-secondary:hover {
  @apply bg-dark-600;
}
</style>