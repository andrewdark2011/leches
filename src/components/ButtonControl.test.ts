import { mount,  } from '@vue/test-utils'
import { TresLeches, useControls } from '/@/'
import { it, expect, describe } from 'vitest'
import { defineComponent, nextTick, ref } from 'vue'
import ButtonControl from './ButtonControl.vue'

describe('Button Controls', () => {
  let wrapper;
  let component;
  const mountComponent = (setup) => {
    component = defineComponent({
      template: `<TresLeches />`,
      components: { TresLeches, ButtonControl },
      setup,
    });

    wrapper = mount(component, {
      components: { TresLeches, ButtonControl },
    });
  };

  it('should render a button', () => {
    mountComponent(() => {
      useControls({acceptBtn: { type: 'button', label: 'Accept' }});
      return
    });
    expect(wrapper.html()).toMatchSnapshot();
    expect(wrapper.find('button').exists()).toBe(true)
  })

  it('should render a button with a label', () => {
    mountComponent(() => {
      useControls({acceptBtn: { type: 'button', label: 'Accept' }});
      return
    });
    expect(wrapper.html()).toMatchSnapshot();
    expect(wrapper.find('button').text()).toBe('Accept')
  })

  it('should render a clickable button', async () => {
    mountComponent(() => {
      const isClicked = ref(false)
      useControls({acceptBtn: { type: 'button', label: 'Accept', onClick: () => { isClicked.value = true } }});
      return { isClicked }
    });

    const button = wrapper.find('button')
    await button.trigger('click')
    expect(wrapper.vm.isClicked).toBe(true)
  })

  it('should render an icon on the button', () => {
    mountComponent(() => {
      useControls({acceptBtn: { type: 'button', label: 'Accept', icon: 'i-carbon-checkmark' }});
      return
    });
    expect(wrapper.html()).toMatchSnapshot();
    expect(wrapper.find('i.i-carbon-checkmark').exists()).toBe(true)
  })

  it('should render a small button', () => {
    mountComponent(() => {
      useControls({acceptBtn: { type: 'button', label: 'Accept', size: 'sm' }});
      return
    });
    expect(wrapper.html()).toMatchSnapshot();
    expect(wrapper.find('button').classes()).toContain('leches-btn-sm')
  })

  it('should render a large button', () => {
    mountComponent(() => {
      useControls({acceptBtn: { type: 'button', label: 'Accept', size: 'lg' }});
      return
    });
    expect(wrapper.html()).toMatchSnapshot();
    expect(wrapper.find('button').classes()).toContain('leches-btn-lg')
  })

  it('should render a block size button', () => {
    mountComponent(() => {
      useControls({acceptBtn: { type: 'button', label: 'Accept', size: 'block' }});
      return
    });
    expect(wrapper.html()).toMatchSnapshot();
    expect(wrapper.find('button').classes()).toContain('leches-btn-block')
  })

  it('should render a primary button variant by default', () => {
    mountComponent(() => {
      useControls({acceptBtn: { type: 'button', label: 'Accept'}});
      return
    });
    expect(wrapper.html()).toMatchSnapshot();
    expect(wrapper.find('button').classes()).toContain('leches-btn-primary')
  })

  it('should render a secondary button variant', () => {
    mountComponent(() => {
      useControls({acceptBtn: { type: 'button', label: 'Accept', variant: 'secondary'}});
      return
    });
    expect(wrapper.html()).toMatchSnapshot();
    expect(wrapper.find('button').classes()).toContain('leches-btn-secondary')
  })
})