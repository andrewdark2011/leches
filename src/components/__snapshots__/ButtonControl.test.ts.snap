// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button Controls > should render a block size button 1`] = `
"<div style=\\"left: 704px; top: 10px;\\" class=\\"absolute select-none z-24 w-280px font-sans text-xs\\">
  <div tabindex=\\"0\\" class=\\"bg-white shadow-xl rounded border-4 border-solid border-black\\">
    <header class=\\"relative cursor-grabbing p-4 flex justify-between text-gray-200 relative\\"><i class=\\"h-4 w-4 p-1.5 flex items-center line-height-0 rounded-full bg-gray-100 text-xs\\">🍰</i>
      <div><i class=\\"i-ic-baseline-drag-indicator\\"></i><i class=\\"i-ic-baseline-drag-indicator\\"></i><i class=\\"i-ic-baseline-drag-indicator\\"></i></div>
      <div></div>
    </header>
    <!--v-if-->
    <div class=\\"p-2 flex\\"><button data-v-9a5f9367=\\"\\" class=\\"leches-btn leches-btn-primary leches-btn-block\\"><i data-v-9a5f9367=\\"\\" class=\\"mr-1\\"></i> Accept</button></div>
  </div>
</div>"
`;

exports[`Button Controls > should render a button 1`] = `
"<div style=\\"left: 704px; top: 10px;\\" class=\\"absolute select-none z-24 w-280px font-sans text-xs\\">
  <div tabindex=\\"0\\" class=\\"bg-white shadow-xl rounded border-4 border-solid border-black\\">
    <header class=\\"relative cursor-grabbing p-4 flex justify-between text-gray-200 relative\\"><i class=\\"h-4 w-4 p-1.5 flex items-center line-height-0 rounded-full bg-gray-100 text-xs\\">🍰</i>
      <div><i class=\\"i-ic-baseline-drag-indicator\\"></i><i class=\\"i-ic-baseline-drag-indicator\\"></i><i class=\\"i-ic-baseline-drag-indicator\\"></i></div>
      <div></div>
    </header>
    <!--v-if-->
    <div class=\\"p-2 inline-flex\\"><button data-v-9a5f9367=\\"\\" class=\\"leches-btn leches-btn-primary leches-btn-sm\\"><i data-v-9a5f9367=\\"\\" class=\\"mr-1\\"></i> Accept</button></div>
  </div>
</div>"
`;

exports[`Button Controls > should render a button with a label 1`] = `
"<div style=\\"left: 704px; top: 10px;\\" class=\\"absolute select-none z-24 w-280px font-sans text-xs\\">
  <div tabindex=\\"0\\" class=\\"bg-white shadow-xl rounded border-4 border-solid border-black\\">
    <header class=\\"relative cursor-grabbing p-4 flex justify-between text-gray-200 relative\\"><i class=\\"h-4 w-4 p-1.5 flex items-center line-height-0 rounded-full bg-gray-100 text-xs\\">🍰</i>
      <div><i class=\\"i-ic-baseline-drag-indicator\\"></i><i class=\\"i-ic-baseline-drag-indicator\\"></i><i class=\\"i-ic-baseline-drag-indicator\\"></i></div>
      <div></div>
    </header>
    <!--v-if-->
    <div class=\\"p-2 inline-flex\\"><button data-v-9a5f9367=\\"\\" class=\\"leches-btn leches-btn-primary leches-btn-sm\\"><i data-v-9a5f9367=\\"\\" class=\\"mr-1\\"></i> Accept</button></div>
  </div>
</div>"
`;

exports[`Button Controls > should render a large button 1`] = `
"<div style=\\"left: 704px; top: 10px;\\" class=\\"absolute select-none z-24 w-280px font-sans text-xs\\">
  <div tabindex=\\"0\\" class=\\"bg-white shadow-xl rounded border-4 border-solid border-black\\">
    <header class=\\"relative cursor-grabbing p-4 flex justify-between text-gray-200 relative\\"><i class=\\"h-4 w-4 p-1.5 flex items-center line-height-0 rounded-full bg-gray-100 text-xs\\">🍰</i>
      <div><i class=\\"i-ic-baseline-drag-indicator\\"></i><i class=\\"i-ic-baseline-drag-indicator\\"></i><i class=\\"i-ic-baseline-drag-indicator\\"></i></div>
      <div></div>
    </header>
    <!--v-if-->
    <div class=\\"p-2 inline-flex\\"><button data-v-9a5f9367=\\"\\" class=\\"leches-btn leches-btn-primary leches-btn-lg\\"><i data-v-9a5f9367=\\"\\" class=\\"mr-1\\"></i> Accept</button></div>
  </div>
</div>"
`;

exports[`Button Controls > should render a primary button variant by default 1`] = `
"<div style=\\"left: 704px; top: 10px;\\" class=\\"absolute select-none z-24 w-280px font-sans text-xs\\">
  <div tabindex=\\"0\\" class=\\"bg-white shadow-xl rounded border-4 border-solid border-black\\">
    <header class=\\"relative cursor-grabbing p-4 flex justify-between text-gray-200 relative\\"><i class=\\"h-4 w-4 p-1.5 flex items-center line-height-0 rounded-full bg-gray-100 text-xs\\">🍰</i>
      <div><i class=\\"i-ic-baseline-drag-indicator\\"></i><i class=\\"i-ic-baseline-drag-indicator\\"></i><i class=\\"i-ic-baseline-drag-indicator\\"></i></div>
      <div></div>
    </header>
    <!--v-if-->
    <div class=\\"p-2 inline-flex\\"><button data-v-9a5f9367=\\"\\" class=\\"leches-btn leches-btn-primary leches-btn-sm\\"><i data-v-9a5f9367=\\"\\" class=\\"mr-1\\"></i> Accept</button></div>
  </div>
</div>"
`;

exports[`Button Controls > should render a secondary button variant 1`] = `
"<div style=\\"left: 704px; top: 10px;\\" class=\\"absolute select-none z-24 w-280px font-sans text-xs\\">
  <div tabindex=\\"0\\" class=\\"bg-white shadow-xl rounded border-4 border-solid border-black\\">
    <header class=\\"relative cursor-grabbing p-4 flex justify-between text-gray-200 relative\\"><i class=\\"h-4 w-4 p-1.5 flex items-center line-height-0 rounded-full bg-gray-100 text-xs\\">🍰</i>
      <div><i class=\\"i-ic-baseline-drag-indicator\\"></i><i class=\\"i-ic-baseline-drag-indicator\\"></i><i class=\\"i-ic-baseline-drag-indicator\\"></i></div>
      <div></div>
    </header>
    <!--v-if-->
    <div class=\\"p-2 inline-flex\\"><button data-v-9a5f9367=\\"\\" class=\\"leches-btn leches-btn-secondary leches-btn-sm\\"><i data-v-9a5f9367=\\"\\" class=\\"mr-1\\"></i> Accept</button></div>
  </div>
</div>"
`;

exports[`Button Controls > should render a small button 1`] = `
"<div style=\\"left: 704px; top: 10px;\\" class=\\"absolute select-none z-24 w-280px font-sans text-xs\\">
  <div tabindex=\\"0\\" class=\\"bg-white shadow-xl rounded border-4 border-solid border-black\\">
    <header class=\\"relative cursor-grabbing p-4 flex justify-between text-gray-200 relative\\"><i class=\\"h-4 w-4 p-1.5 flex items-center line-height-0 rounded-full bg-gray-100 text-xs\\">🍰</i>
      <div><i class=\\"i-ic-baseline-drag-indicator\\"></i><i class=\\"i-ic-baseline-drag-indicator\\"></i><i class=\\"i-ic-baseline-drag-indicator\\"></i></div>
      <div></div>
    </header>
    <!--v-if-->
    <div class=\\"p-2 inline-flex\\"><button data-v-9a5f9367=\\"\\" class=\\"leches-btn leches-btn-primary leches-btn-sm\\"><i data-v-9a5f9367=\\"\\" class=\\"mr-1\\"></i> Accept</button></div>
  </div>
</div>"
`;

exports[`Button Controls > should render an icon on the button 1`] = `
"<div style=\\"left: 704px; top: 10px;\\" class=\\"absolute select-none z-24 w-280px font-sans text-xs\\">
  <div tabindex=\\"0\\" class=\\"bg-white shadow-xl rounded border-4 border-solid border-black\\">
    <header class=\\"relative cursor-grabbing p-4 flex justify-between text-gray-200 relative\\"><i class=\\"h-4 w-4 p-1.5 flex items-center line-height-0 rounded-full bg-gray-100 text-xs\\">🍰</i>
      <div><i class=\\"i-ic-baseline-drag-indicator\\"></i><i class=\\"i-ic-baseline-drag-indicator\\"></i><i class=\\"i-ic-baseline-drag-indicator\\"></i></div>
      <div></div>
    </header>
    <!--v-if-->
    <div class=\\"p-2 inline-flex\\"><button data-v-9a5f9367=\\"\\" class=\\"leches-btn leches-btn-primary leches-btn-sm\\"><i data-v-9a5f9367=\\"\\" class=\\"i-carbon-checkmark mr-1\\"></i> Accept</button></div>
  </div>
</div>"
`;
